app/mattermost.js

```diff
@@ -6,14 +6,38 @@ import React from 'react';
 import {Provider} from 'react-redux';
 
 import Root from 'app/components/root';
-import initialState from 'app/initial_state';
 import Router from 'app/navigation/router';
 import configureStore from 'app/store';
 
+import initialState from './initial_state';
+
 const store = configureStore(initialState);
 
 export default class Mattermost extends React.Component {
+    state = {
+        hydrated: false
+    }
+
+    componentWillMount() {
+        this.unsubscribeFromStore = store.subscribe(this.listenForHydration);
+    }
+
+    // We need to wait for hydration to occur before load the router.
+    listenForHydration = () => {
+        const state = store.getState();
+        if (state.views.root.hydrationComplete) {
+            this.unsubscribeFromStore();
```

Comment:
The hydration is kind of finicky. We don't want to run any actions that may modify the state before hydration is complete. I felt like the best way to accomplish that is to make sure that we don't load anything until hydration is complete. So this just basically makes sure that the store won't dispatch any actions until hydration is complete.